formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/ProtoTypes.kt

```diff
@@ -29,18 +37,48 @@ public enum class ProtoNumberType {
 @Target(AnnotationTarget.PROPERTY)
 public annotation class ProtoType(public val type: ProtoNumberType)
 
-internal typealias ProtoDesc = Pair<Int, ProtoNumberType>
+internal typealias ProtoDesc = Long
 
-internal fun extractParameters(descriptor: SerialDescriptor, index: Int, zeroBasedDefault: Boolean = false): ProtoDesc {
-    val protoId = descriptor.findAnnotation<ProtoId>(index)
-    val idx = protoId?.id ?: (if (zeroBasedDefault) index else index + 1)
-    val format = descriptor.findAnnotation<ProtoType>(index)?.type
-            ?: ProtoNumberType.DEFAULT
-    return idx to format
+@Suppress("NOTHING_TO_INLINE")
+internal inline fun ProtoDesc(protoId: Int, type: ProtoNumberType): ProtoDesc {
+    return type.signature or protoId.toLong()
 }
 
-public class ProtobufDecodingException(message: String) : SerializationException(message)
+internal inline val ProtoDesc.protoId: Int get() = (this and Int.MAX_VALUE.toLong()).toInt()
+
+internal val ProtoDesc.numberType: ProtoNumberType get() {
+    if (ProtoNumberType.DEFAULT.equalTo(this)) return ProtoNumberType.DEFAULT
+    if (ProtoNumberType.SIGNED.equalTo(this)) return ProtoNumberType.SIGNED
+    return ProtoNumberType.FIXED
+}
 
-internal inline fun <reified A: Annotation> SerialDescriptor.findAnnotation(elementIndex: Int): A? {
-    return getElementAnnotations(elementIndex).find { it is A } as A?
+internal fun SerialDescriptor.extractParameters(index: Int): ProtoDesc {
+    val annotations = getElementAnnotations(index)
+    var protoId: Int = index + 1
+    var format: ProtoNumberType = ProtoNumberType.DEFAULT
+    for (i in annotations.indices) { // Allocation-friendly loop
```

Comment:
why it is more allocation-friendly than `for annotation in annotations`?

Reply:
It doesn't allocate an iterator